FirebaseStorageUITests/FUIImageViewCategoryTests.m

```diff
@@ -0,0 +1,118 @@
+@property (nonatomic, readwrite) UIImageView *imageView;
+@property (nonatomic, readwrite) SDImageCache *cache;
+@end
+
+@implementation FUIImageViewCategoryTests
+
+- (void)setUp {
+  [super setUp];
+  self.ref = OCMClassMock([FIRStorageReference class]);
+  OCMStub([self.ref fullPath]).andReturn(@"path/to/image.png");
+  self.cache = OCMClassMock([SDImageCache class]);
+  self.imageView = [[UIImageView alloc] initWithFrame:CGRectMake(0, 0, 100, 100)];
+}
+
+- (void)tearDown {
+  [super tearDown];
+}
+
+- (void)testItCreatesADownloadTaskIfCacheIsEmpty {
+  OCMStub([self.ref dataWithMaxSize:512 completion:[OCMArg any]])
+    .andReturn(OCMClassMock([FIRStorageDownloadTask class]));
+  FIRStorageDownloadTask *download = [self.imageView sd_setImageWithStorageReference:self.ref
+                                                                        maxImageSize:512
+                                                                    placeholderImage:nil
+                                                                               cache:self.cache
+                                                                          completion:^(UIImage *image,
+                                                                                       NSError *error,
+                                                                                       SDImageCacheType cacheType,
+                                                                                       FIRStorageReference * storageRef) {
+    XCTAssert(self.imageView.image == image, @"expected download to populate image");
+    XCTAssertNil(error, @"expected successful download to not produce an error");
+  }];
+  XCTAssertNotNil(download, @"expected image view with empty cache to attempt a download");
+}
+
+- (void)testItDoesNotCreateADownloadIfImageIsCached {
+  OCMStub([self.ref dataWithMaxSize:4096 completion:[OCMArg any]])
+    .andReturn(OCMClassMock([FIRStorageDownloadTask class]));
+  UIImage *image = [[UIImage alloc] init];
+  self.cache = [[SDImageCache alloc] initWithNamespace:@"FirebaseStorageUITests"];
+  [self.cache storeImage:image forKey:self.ref.fullPath];
+  FIRStorageDownloadTask *download = [self.imageView sd_setImageWithStorageReference:self.ref
+                                                                        maxImageSize:4096
+                                                                    placeholderImage:nil
+                                                                               cache:self.cache
+                                                                          completion:nil];
+  XCTAssertNil(download, @"expected image view to not create new download when fetching cached image");
+  XCTAssert(self.imageView.image == image, @"expected image view to use cached image");
+  [self.cache cleanDisk];
+  [self.cache clearMemory];
+}
+
+- (void)testItRaisesAnErrorIfDownloadingFails {
+  [self.imageView sd_setImageWithStorageReference:self.ref
```

Comment:
BTW, happy if we don't (it was messy), just surprised that we don't (or maybe we do and the tests are flaky?)